openshift/machine-config-operator
# MCO-1202: MCO-1203: MCO-1204: MCO-1205: MCO-1213: Implementing tlsSecurityProfile for MCO

**- What I did**

In regular cluster operation:
- The operator and controller will listen on an APIServer object, which contains the [global cluster tls settings](https://docs.openshift.com/container-platform/4.15/security/tls-security-profiles.html#tls-profiles-kubernetes-configuring_tls-security-profiles).
- The operator will now render TLSMinVersion and TLSCiphers to the MCC and MCD kube-rbac-proxy sidecar manifests via templated arguments.
- The template controller will now render TLSMinVersion and TLSCiphers to the kube-rbac-proxy-crio metrics pods via templated arguments. These are per node pods in the MCO namespace, and deployed via MachineConfigs. As a result changing the tls settings will cause a MachineConfig rollout.
- The MCS container now has two new TLS arguments. The operator will render TLSMinVersion and TLSCiphers via these new arguments in the MCS daemonset manifest. The MCS will attempt to create a http server with these new TLS settings while starting up.

During cluster bootstrap:
- The bootstrap MCC will default to the intermediate security profile, unless an install time APIServer manifest is provided. This is necessary during bootstrap, as the kube-rbac-proxy-crio pod manifests show up in rendered MachineConfigs and we need the bootstrap MachineConfigs to match the in cluster MachineConfigs post install. 
- The bootstrap MCC has access to all install time manifests provided to the installer, but the bootstrap MCS does not. When the bootstrap MCC determines that an APIServer manifest has been provided, it will write it to bootstrap MCS's directory.
- The bootstrap MCS will now read in the APIServer manifest and launch its http server with the TLS settings defined in the manifest. If no manifest is provided, it will default to the intermediate security profile. 

**- How to verify it**
1. Bring up a cluster with an install time APIServer manifest named "cluster". To test bootstrap behavior, ensure
environment variable `OPENSHIFT_INSTALL_PRESERVE_BOOTSTRAP` is set to true.
3. Once the install is complete, observe the kube rbac proxy manifests in cluster to ensure they have the correct TLS configuration. These can be seen as arguments for the kube rbac proxy sidecards on the MCO deployments, as well as encapsulated in MachineConfigs for the kube-rbac-proxy-crio pods.
4. The MCS daemonset will log the current TLS settings at startup. You can also test the MCS endpoint using the method described in this [comment](https://issues.redhat.com/browse/MCO-1204?focusedId=25007192&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-25007192).
5. You can also log in to the bootstrap node and verify that the MCS bootstrap logs started with the correct TLS settings.
6.  Switch between TLS profiles by [following the documentation](https://docs.openshift.com/container-platform/4.15/security/tls-security-profiles.html#tls-profiles-kubernetes-configuring_tls-security-profiles).
7. This will cause a new MachineConfig rollout and also cause the MCC, MCD and MCS pods to restart. You can verify (2) and (3) is as expected as the update rolls through the cluster.

**Things to note**: 

- It may take a few moments before the operator rolls out the new manifests. If it is done while the MCO is doing a MachineConfig update to the master pool, the manifests won't be updated until the master pool is finished updating. This is because of the way the operator's sync loops are structured.
- The APIServer rejects the Modern profile at the moment, according to the docs.
- If no tlsSecurityProfile is provided, the MCO will default to Intermediate.
